examples/step-27/step-27.cc

```diff
@@ -612,141 +685,63 @@ namespace Step27
 
   // @sect4{LaplaceProblem::estimate_smoothness}
 
+  // As described in the introduction, we will need to take the maximum
+  // absolute value of fourier coefficients which correspond to $k$-vector
+  // $|{\bf k}|= const$. To filter the coefficients Table we
+  // will use the FESeries::process_coefficients() which requires a predicate
+  // to be specified. The predicate should operate on TableIndices and return
+  // a pair of <code>bool</code> and <code>unsigned int</code>. The latter
+  // is the value of the map from TableIndicies to unsigned int.  It is
+  // be used to define subsets coefficients from which we search for the one
+  // with highest absolute value, i.e. $l^\infty$-norm. Whereas <code>bool</code>
+  // parameter defiens which indices should be used in processing. In the
+  // current case we are interested in coefficients which correspond to
+  // $0 < i*i+j*j < N*N$ and $0 < i*i+j*j+k*k < N*N$ in 2D and 3D, respectively.
+  template <int dim>
+  std::pair<bool,unsigned int>
+  predicate_ind(const TableIndices<dim> &ind);
+
+  template<>
+  std::pair<bool,unsigned int>
+  predicate_ind<2>(const TableIndices<2> &ind)
+  {
+    const unsigned int v = ind[0]*ind[0]+ind[1]*ind[1];
+    if (v>0 && v<7*7)
+      return std::make_pair(true,v);
+    else
+      return std::make_pair(false,v);
+  }
+
+  template<>
+  std::pair<bool,unsigned int>
+  predicate_ind<3>(const TableIndices<3> &ind)
+  {
+    const unsigned int v = ind[0]*ind[0]+ind[1]*ind[1]+ind[2]*ind[2];
+    if (v>0 && v < 5*5*5)
+      return std::make_pair(true,v);
+    else
+      return std::make_pair(false,v);
+  }
```

Comment:
correct. Perhaps one should wrap it into the class function and to `std::bind()` instead.